src/document.js

```diff
@@ -600,6 +616,23 @@ class DocumentSnapshot {
     };
   }
 
+  /**
+   * Returns true if the document's data and path in this `DocumentSnapshot` is
+   * equal to the provided value.
+   *
+   * @param {*} other The value to compare against.
+   * @return {boolean} true if this `DocumentSnapshot` is equal to the provided
+   * value.
+   */
+  isEqual(other) {
+    return (
+      this === other ||
+      (is.instance(other, DocumentSnapshot) &&
+        this._ref.isEqual(other._ref) &&
+        deepEqual(this._fieldsProto, other._fieldsProto))
```

Comment:
I updated the code here and in the Query comparison to use strict equality. Note that the comparison here uses the underlying Protobuf representation and hence '3' is already different from 3. I agree though that it does make sense to not rely on this here, since I had to verify this myself first.